a10networks/a10-octavia
# Feature/write memory enhancement

## Description
Write Memory Enhancement feature 

https://teams.microsoft.com/_?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f#/docx/viewer/teams/https:~2F~2Fa10networks.sharepoint.com~2Fsites~2FOpenstack~2FShared%20Documents~2FDevelopment~2FResearch%20&%20Design~2FHouse%20Keeper%20Write%20Memory%20Feature~2Fa10_house_keeper_write_mem.docx?threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&baseUrl=https:~2F~2Fa10networks.sharepoint.com~2Fsites~2FOpenstack&fileId=4bf74aca-8c9f-4669-9f62-7e5949e0c1aa&ctx=files&rootContext=items_view&viewerAction=view


## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1908

## Technical Approach
```
- Added config support in `[a10_house_keeping]` section like `use_periodic_write_memory` and `write_mem_interval`

- Housekeeper will fetch "eligible" thunders after every configured write_mem_interval of time from vthunder db

- Thunders are filtered again based on unique ip_address and partition_name

- Loop through thunder list and perform 3 tasks:
    -> Set associated LBs to PENDING_UPDATE
    -> Write memory api call in shared partition
    -> Write memory api call in specified partition (based on vthunder.partition_name)
    -> Set associated LBs to ACTIVE 
  
- If `WriteMemoryHouseKeeper` task fails, `updated_at` is set as current time for that thunder and flow contrinues.
- If `use_periodic_write_memory` is `disable`, then default controller worker `WriteMemory` calls are made.
- Added new `SetThunderUpdatedAt` task at end of each flow in `lb, listener, hm, member, pool, l7policy, l7rule flows.(create,update and delete)`
- Placed more logs in tasks
- Added corresponding UTs.
```

## Config Changes
<pre>
[a10_house_keeping]

<b>use_periodic_write_memory = 'enable'
write_mem_interval = 300</b>
</pre>

## Test Cases
1. If `use_periodic_write_memory` is `disable` (default), then default controller-worker write memory will take place.
2. if  `use_periodic_write_memory` is `enable', then write memory feature is enabled.
3. If `write_mem_interval` set less than 300, will get error in house-keeper service.
![image](https://user-images.githubusercontent.com/52992745/103641961-86baf000-4f78-11eb-89a6-d98f8cb84764.png)

4. If `write_mem_interval` is not set, then default value `3600` is set.
5. If `write_mem_interval` is set to some value (>=300), then thunders that are _**recently updated**_ within last `write_mem_interval` time, is picked for write_memory operation

**_(recently updated using openstack create/update/delete of any (LB, listener, pool, hm, member, l7policy, l7rule) )_**
   
6. With above scenario, if exception occur in write memory, then simple `updated_at` field is set which makes it eligible for write_memory in further threads.
7. If some exception occurs, LOG is printed and thread continues.
READTIME out exception: 
![image](https://user-images.githubusercontent.com/52992745/103640283-bfa59580-4f75-11eb-81c1-f04fa0604282.png)
![image](https://user-images.githubusercontent.com/52992745/103640384-eb288000-4f75-11eb-8a98-bafbce9dea28.png)

8. If no vthunders are recently updated within before `write_mem_interval`, then warning is printed
![image](https://user-images.githubusercontent.com/52992745/103641313-938b1400-4f77-11eb-8312-432a1befff43.png)



## Manual Testing
STEP 1 : Enable Write mem in housekeeper by setting `use_periodic_write_memory` as `enable` and set a `write_mem_interval` as some seconds
STEP 2: Restart `a10-house-keeper.service`
![image](https://user-images.githubusercontent.com/52992745/103640006-46a63e00-4f75-11eb-901f-5b83d5eec8e8.png)

STEP 3: Create a LB, (other resources associated with it).
STEP 4: Check LOGS on `sudo journalctl -af --unit a10-house-keeper.service`. You will find that Write memory performed on shared and specific partition.
![image](https://user-images.githubusercontent.com/52992745/103641136-46a73d80-4f77-11eb-946f-10d0a448514f.png)
STEP 5: Check `show audit` to see write memory apis beings called on vthunders.
![image](https://user-images.githubusercontent.com/52992745/103641682-2166ff00-4f78-11eb-87c7-a8e47c248c49.png)
